### PR TITLE
WFLY-4621 Sync BouncyCastle version with WildFly

### PR DESCRIPTION
**Pull request by liweinan**

### CODE REVIEW

File: jaxrs/security/resteasy-crypto/pom.xml [hunk 1]
```diff
@@ -52,15 +52,16 @@
...
            <groupId>org.bouncycastle</groupId>
-           <artifactId>bcmail-jdk16</artifactId>
+           <artifactId>bcmail-jdk15on</artifactId>
         </dependency>
         <dependency>
-            <groupId>javax.mail</groupId>
-            <artifactId>mail</artifactId>
+            <groupId>com.sun.mail</groupId>
+            <artifactId>javax.mail</artifactId>
+            <version>1.5.3</version>
```
Comment (ctomc) [inline jaxrs/security/resteasy-crypto/pom.xml]: no need for version here it is defined in parent pom, otherwise use ${dep.com.sun.mail.version}

Comment (liweinan) [inline jaxrs/security/resteasy-crypto/pom.xml]: Thanks for checking Tomaz :-) version removed.

**Comment by ronsigal** [comment]
Hey Wei,

I've run into a couple of problems.

* In order to build the zip files in jboss-modules, I had to change the lines

            <maven-resource group="org.bouncycastle" artifact="bcprov-jdk16"/>
            <maven-resource group="org.bouncycastle" artifact="bcmail-jdk16"/>

in build.xml and build-wf8.xml.

* When I ran the arquillian tests, I saw

22:06:08,357 WARN  [org.jboss.resteasy.plugins.providers.RegisterBuiltin] (default task-20) NoClassDefFoundError: Unable to load builtin provider: org.jboss.resteasy.security.smime.EnvelopedWriter
22:06:08,357 WARN  [org.jboss.resteasy.plugins.providers.RegisterBuiltin] (default task-20) NoClassDefFoundError: Unable to load builtin provider: org.jboss.resteasy.security.smime.MultipartSignedReader
22:06:08,358 WARN  [org.jboss.resteasy.plugins.providers.RegisterBuiltin] (default task-20) NoClassDefFoundError: Unable to load builtin provider: org.jboss.resteasy.security.smime.PKCS7SignatureWriter

in the wildfly 8 tests. I added 

        <dependency>
            <groupId>org.bouncycastle</groupId>
            <artifactId>bcpkix-jdk15on</artifactId>
        </dependency>

to jboss-modules/pom.xml and

            <maven-resource group="org.bouncycastle" artifact="bcpkix-jdk15on"/>

to the org.bouncycastle module, and then I just saw

22:31:24,235 WARN  [org.jboss.resteasy.plugins.providers.RegisterBuiltin] (MSC service thread 1-12) NoClassDefFoundError: Unable to load builtin provider: org.jboss.resteasy.security.smime.MultipartSignedReader

I don't know why it's not finding MultipartSignedReader ...

**Comment by liweinan** [comment]
Hi Ron,

Sorry I've forgotten to tell you that you need this patch to deploy to wildfly: https://github.com/wildfly/wildfly/pull/7458/files

Could you please try again?

**Comment by ronsigal** [comment]
Ah, I'll bet that's it. I got

20:55:05,635 WARN  [org.jboss.resteasy.plugins.providers.RegisterBuiltin] (default task-2) NoClassDefFoundError: Unable to load builtin provider: org.jboss.resteasy.security.smime.MultipartSignedReader
20:55:05,636 ERROR [stderr] (default task-2) java.lang.NoClassDefFoundError: javax/activation/DataSource

when RegisterBuiltin tried to load MultipartSignedReader.

**Comment by ronsigal** [comment]
But we've got org.bouncycastle in our jboss-modules zip files. Shouldn't we include 

   <module name="javax.mail.api" optional="true"/>
   <module name="javax.activation.api" optional="true"/>

as well?

**Comment by ronsigal** [comment]
Aaarrgghh!! I've got

<module xmlns="urn:jboss:module:1.1" name="org.bouncycastle">
    <resources>
        <resource-root path="bcprov-jdk15on-1.52.jar"/>
        <resource-root path="bcmail-jdk15on-1.52.jar"/>
        <resource-root path="bcpkix-jdk15on-1.52.jar"/>
        
    </resources>
    <dependencies>
        <module name="javax.api"/>
        <module name="javax.mail.api" optional="true"/>
        <module name="javax.activation.api" optional="true"/>
    </dependencies>

</module>

and I'm still getting that ClassNotFoundException.
in

**Comment by ronsigal** [comment]
Ok, if I use wildfly-10.0.0.Alpha4 out of the box, there's no problem. But if I unzip resteasy-jboss-modules-wf8-3.0.12.Final-SNAPSHOT.zip, then I get the NoClassDefFoundError.

**Comment by ronsigal** [comment]
It turns out that wildfly-10.0.0.Alpha4 out of the box was not throwing an exception because the resteasy-crypto services weren't being loaded. When I refered to MultipartSignedReader in a test resource, I got the exception. Finally, I made it go away by adding

        <module name="javax.activation.api"/>

to the resteasy-crypto module. I want to run the basic integration tests in WF8, WF9, and WF10 with the resteasy zipped modules and see what happens.

But I've got another problem. When I build jboss-modules with jdk 1.8, I get

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.7:run (build-dist-as7) on project resteasy-jboss-modules: An Ant BuildException has occured: The following error occurred while executing this line:
[ERROR] /home/rsigal/tmp/workspace3/Resteasy/jaxrs/jboss-modules/build.xml:34: The following error occurred while executing this line:
[ERROR] /home/rsigal/tmp/workspace3/Resteasy/jaxrs/jboss-modules/build.xml:55: The following error occurred while executing this line:
[ERROR] /home/rsigal/tmp/workspace3/Resteasy/jaxrs/jboss-modules/build.xml:56: The following error occurred while executing this line:
[ERROR] /home/rsigal/tmp/workspace3/Resteasy/jaxrs/jboss-modules/lib.xml:172: javax.script.ScriptException: ReferenceError: "importClass" is not defined in <eval> at line number 8
[ERROR] around Ant part ...<ant antfile="build.xml" inheritRefs="true">... @ 4:47 in /home/rsigal/tmp/workspace3/Resteasy/jaxrs/jboss-modules/target/antrun/build-main.xml
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

Strange.

**Comment by ronsigal** [comment]
With the patch

diff --git a/jaxrs/jboss-modules/build-wf8.xml b/jaxrs/jboss-modules/build-wf8.xml
index 58314be..751ff64 100755
--- a/jaxrs/jboss-modules/build-wf8.xml
+++ b/jaxrs/jboss-modules/build-wf8.xml
@@ -53,8 +53,9 @@
           -->
 
         <module-def name="org.bouncycastle">
-            <maven-resource group="org.bouncycastle" artifact="bcprov-jdk16"/>
-            <maven-resource group="org.bouncycastle" artifact="bcmail-jdk16"/>
+            <maven-resource group="org.bouncycastle" artifact="bcprov-jdk15on"/>
+            <maven-resource group="org.bouncycastle" artifact="bcmail-jdk15on"/>
+            <maven-resource group="org.bouncycastle" artifact="bcpkix-jdk15on"/>
         </module-def>
 
         <module-def name="org.codehaus.jackson.jackson-core-asl">
diff --git a/jaxrs/jboss-modules/build.xml b/jaxrs/jboss-modules/build.xml
index 810a002..276b135 100755
--- a/jaxrs/jboss-modules/build.xml
+++ b/jaxrs/jboss-modules/build.xml
@@ -53,8 +53,8 @@
           -->
 
         <module-def name="org.bouncycastle">
-            <maven-resource group="org.bouncycastle" artifact="bcprov-jdk16"/>
-            <maven-resource group="org.bouncycastle" artifact="bcmail-jdk16"/>
+            <maven-resource group="org.bouncycastle" artifact="bcprov-jdk15on"/>
+            <maven-resource group="org.bouncycastle" artifact="bcmail-jdk15on"/>
         </module-def>
 
         <module-def name="org.codehaus.jackson.jackson-core-asl">
diff --git a/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml b/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml
index 87d8c8e..46c0996 100644
--- a/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml
+++ b/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml
@@ -35,5 +35,6 @@
         <module name="javax.ws.rs.api"/>
         <module name="org.jboss.resteasy.resteasy-jaxrs"/>
         <module name="org.bouncycastle" export="true"/>
+        <module name="javax.activation.api"/>
     </dependencies>
 </module>

I get the basic integration testsuite for WF 8, 9, and 10 with resteasy-jboss-modules-wf8-3.0.12.Final-SNAPSHOT.zip unzipped.

**Comment by ronsigal** [comment]
Also, I ran arquillian tests for WF8, 9, and 10, and didn't see any ClassNotFoundExceptions

**Comment by liweinan** [comment]
Thanks for detailed checking Ron! Could you please provide me a full patch for this one so I can update this PR?

**Comment by ronsigal** [comment]
Wei, I've attached two versions of a patch to WFLY-4621. Could you create a RESTEASY JIRA issue, since we're making changes to Resteasy?

**Comment by liweinan** [comment]
Thanks Ron!  RESTEasy JIRA Issue created: https://issues.jboss.org/browse/RESTEASY-1195

**Comment by ronsigal** [comment]
Thanks Wei!  ;-)

**Comment by liweinan** [comment]
Hi Ron,

I'm confused about this addition:

```
diff --git a/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml b/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml
index 87d8c8e..46c0996 100644
--- a/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml
+++ b/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/module.xml
@@ -35,5 +35,6 @@
         <module name="javax.ws.rs.api"/>
         <module name="org.jboss.resteasy.resteasy-jaxrs"/>
         <module name="org.bouncycastle" export="true"/>
+        <module name="javax.activation.api"/>
     </dependencies>
 </module>
```

Because we already patched in bouncycastle module:

```
feature-pack/src/main/resources/modules/system/layers/base/org/bouncycastle/main/module.xml
 @@ -30,6 +30,8 @@
     </resources>
     <dependencies>
         <module name="javax.api"/>
+        <module name="javax.mail.api" optional="true"/>
+        <module name="javax.activation.api" optional="true"/>
     </dependencies>
 
 </module>
```

Will it be necessary to include activation again in resteasy-crypto? Sorry maybe I'm too blind on wildfly module internals. Could you please help to confirm the addition of javax.activation.api in resteasy-crypto?

Or maybe we can also mark it as optional="true"?  Because @bstansberry commented in https://github.com/wildfly/wildfly/pull/7458 :

```
bstansberry commented on May 14
I don't want to block this PR over this, but I wonder if we need a different approach to module dependencies for stuff like this. This change basically means if someone needs bouncycastle, they also must have javax.mail in their installation, at a cost of 579K, even if they have no interest in this particular functionality.

Is optional=true the way to handle this kind of thing? That provides no clue to the user as to what functionality they lose by not adding the module (although in this case it's easy to guess it's related to mail) but it does allow them to create a server that doesn't require stuff they don't plan to use.
```

**Comment by liweinan** [comment]
@ronsigal btw, I always flooded by resteasy notifications after Bill grant me the upstream merge permissions. And I may miss your important comments and PRs you'd like me to work on. If you send me some important things in github and I didn't response, please send to me email directly and I'll catch it :-)

**Comment by ronsigal** [comment]
Hmmm, it could be a change I made when I was trying to fix the missing 
providers problem, and maybe it's not necessary. I'll see what happens 
when I take it out.

-Ron

**Comment by ronsigal** [comment]
When I take it out, I go back to getting

>23:07:23,013 WARN  [org.jboss.resteasy.plugins.providers.RegisterBuiltin] (default task-1) NoClassDefFoundError: Unable to load builtin provider: org.jboss.resteasy.security.smime.MultipartSignedReader

It looks like an alternative is to add 'export="true"' in org/bouncycastle/main/module.xml:

>&lt;module name="javax.activation.api" optional="true" export="true"/&gt;

I guess Wildfly would have to be changed as well?

**Comment by ronsigal** [comment]
re: "If you send me some important things in github and I didn't response, please send to me email directly"

Good to know. ;-)

**Comment by liweinan** [comment]
Thanks for detail checking Ron :-) Let me discuss with Brian whether we put 'export=true' in bouncycastle or add 'javax.activation.api' in resteasy-crypto

**Comment by liweinan** [comment]
@ronsigal  I have updated this PR. Could you please help to double check before I merge and close? Thank you very much! :-)

**Comment by liweinan** [comment]
@ronsigal There is only one line different with WildFly upstream now:
resteasy-upstream/jaxrs/jboss-modules/src/main/resources/modules/org/jboss/resteasy/resteasy-crypto/main/modules.xml

```
        <module name="org.bouncycastle" export="true"/>
```

In wildfly it's:
/wildfly-upstream/feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-crypto/main/modules.xml

```
        <module name="org.bouncycastle"/>
```

Do you think export org.bouncycastle is necessary? If so, I'll open another PR to wildfly to sync with our side. Or vice versa if not necessary.

**Comment by ronsigal** [comment]
Hey Wei,

I've been looking at resteasy-crypto in wildfly, and it turns out that

|        <module name="javax.activation.api" optional="true"/>

|

has been there going back to WF 8. Somehow, it just got left out in 
Resteasy's jboss-modules.

-Ron

**Comment by ronsigal** [comment]
I've tested the new version of the pull request

 * Resteasy testsuite, including the arquillian tests
 * temporary arquillian tests for WF9 and WF10
 * basic integration testsuite in WF8, WF9, and WF10

Everything passes!!

**Comment by ronsigal** [comment]
re:  &lt;module name="org.bouncycastle" export="true"/&gt;

I don't know why we have it exported in jboss-modules. Nothing in the tests I just ran suggest that it's necessary.

**Comment by liweinan** [comment]
Thanks for great help on this Ron! 

MERGED :-)

**Comment by ronsigal** [comment]
Excellent!! ;-)

**Pull request merged**